{"cells": [{"cell_type": "markdown", "metadata": {}, "source": ["# Classification and Prediction in GenePattern Notebook\n", "\n", "This notebook will show you how to use k-Nearest Neighbors (kNN) to build a predictor, use it to classify human airway smooth muscle as treated or untreated by asthma medications, and assess its accuracy in cross-validation."]}, {"cell_type": "markdown", "metadata": {}, "source": ["### K-nearest-neighbors (KNN)\n", "KNN classifies an unknown sample by assigning it the phenotype label most frequently represented among the k nearest known samples (Golub and Slonim et al., 1999). \n", "\n", "Additionally, you can select a weighting factor for the 'votes' of the nearest neighbors. For example, one might weight the votes by the reciprocal of the distance between neighbors to give closer neighors a greater vote."]}, {"cell_type": "markdown", "metadata": {}, "source": ["<h2>1. Log in to GenePattern</h2>\n", "\n", "<ul>\n", "\t<li>Select Broad Institute as the server</li>\n", "\t<li>Enter your username and password.</li>\n", "\t<li>Click <em>Login to GenePattern</em>.</li>\n", "\t<li>When you are logged in, you can click the - button in the upper right hand corner to collapse the cell.</li>\n", "\t<li>Alternatively, if you are prompted to Login as your username, just click that button and give it a couple seconds to authenticate.</li>\n", "</ul>\n"]}, {"cell_type": "code", "execution_count": 4, "metadata": {"genepattern": {"name": "Login", "server": "https://cloud.genepattern.org/gp", "type": "auth"}}, "outputs": [{"data": {"application/vnd.jupyter.widget-view+json": {"model_id": "4175b0ffefcc4031bbf5a7047a07e24f", "version_major": 2, "version_minor": 0}, "text/plain": ["GPAuthWidget()"]}, "metadata": {}, "output_type": "display_data"}], "source": ["# Requires GenePattern Notebook: pip install genepattern-notebook\n", "import gp\n", "import genepattern\n", "\n", "# Username and password removed for security reasons.\n", "genepattern.display(genepattern.session.register(\"https://cloud.genepattern.org/gp\", \"\", \"\"))"]}, {"cell_type": "markdown", "metadata": {}, "source": ["<h2 id=\"2.-Preprocess-raw-count-data\">2. Preprocess raw count data</h2>\n", "\n", "<ul>\n", "\t<li>Preprocess RNA-Seq count data in a GCT file so that it is suitable for use in GenePattern analyses.</li>\n", "</ul>\n", "\n", "<div class=\"alert alert-info\">\n", "<p>For the <strong><em>input.file</em></strong> parameter, click &quot;Add Path or URL...&quot; then copy and paste this URL into the <em>&quot;Enter Path or URL&quot; </em>text box, and click <strong><em>Select</em></strong>:</p>\n", "\n", "<p><a href=\"https://datasets.genepattern.org/data/VIB/MergedHTSeqCounts_GSE52778.gct\" target=\"_blank\">https://datasets.genepattern.org/data/VIB/MergedHTSeqCounts_GSE52778.gct</a></p>\n", "\n", "<p>&nbsp;</p>\n", "\n", "<p>For the <strong><em>cls.file</em></strong> parameter, click &quot;Add Path or URL...&quot; then copy and paste this URL into the <em>&quot;Enter Path or URL&quot; </em>text box, and click <strong><em>Select</em></strong>:</p>\n", "\n", "<p><a href=\"https://datasets.genepattern.org/data/VIB/MergedHTSeqCounts_GSE52778.cls\" target=\"_blank\">https://datasets.genepattern.org/data/VIB/MergedHTSeqCounts_GSE52778.cls</a></p>\n", "\n", "<p>&nbsp;</p>\n", "\n", "<p>Click the button <strong><em>Run</em></strong> on the analysis below.</p>\n", "</div>\n"]}, {"cell_type": "code", "execution_count": 5, "metadata": {"genepattern": {"param_values": {"cls.file": [], "expression.value.filter.threshold": null, "input.file": [], "job.cpuCount": null, "job.memory": null, "job.walltime": null, "output.file": null}, "type": "task"}}, "outputs": [{"data": {"application/vnd.jupyter.widget-view+json": {"model_id": "05b9f96264fb4db6af8ae5472a39d0ef", "version_major": 2, "version_minor": 0}, "text/plain": ["GPTaskWidget(lsid='urn:lsid:broad.mit.edu:cancer.software.genepattern.module.analysis:00355')"]}, "metadata": {}, "output_type": "display_data"}], "source": ["preprocessreadcounts_task = gp.GPTask(genepattern.get_session(0), 'urn:lsid:broad.mit.edu:cancer.software.genepattern.module.analysis:00355')\n", "preprocessreadcounts_job_spec = preprocessreadcounts_task.make_job_spec()\n", "preprocessreadcounts_job_spec.set_parameter(\"input.file\", \"\")\n", "preprocessreadcounts_job_spec.set_parameter(\"cls.file\", \"\")\n", "preprocessreadcounts_job_spec.set_parameter(\"output.file\", \"<input.file_basename>.preprocessed.gct\")\n", "preprocessreadcounts_job_spec.set_parameter(\"expression.value.filter.threshold\", \"1\")\n", "genepattern.display(preprocessreadcounts_task)"]}, {"cell_type": "markdown", "metadata": {}, "source": ["<h2 id=\"3.-Run-k-Nearest-Neighbors-Cross-Validation\">3. Run k-Nearest Neighbors Cross Validation</h2>\n", "\n", "<p>In the result cell for the PreprocessReadCounts job, you will see 2 files.</p>\n", "\n", "<div class=\"alert alert-info\">Click the &quot;i&quot; icon next to the MergedHTSeqCounts_GSE52778.preprocessed.gct file.</div>\n", "\n", "<p>You will see a dialog box with several options.</p>\n", "\n", "<div class=\"alert alert-info\">\n", "<ul>\n", "\t<li>Select &quot;Send to existing GenePattern Cell&quot;</li>\n", "\t<li>Choose &quot;KNNXvalidation&quot;</li>\n", "\t<li>Click Run.</li>\n", "</ul>\n", "</div>"]}, {"cell_type": "code", "execution_count": 6, "metadata": {"genepattern": {"param_values": {"class.filename": null, "data.filename": [], "distance.measure": null, "feature.selection.statistic": null, "feature.summary.file": null, "job.cpuCount": null, "job.memory": null, "job.walltime": null, "min.std": null, "num.features": null, "num.neighbors": null, "pred.results.file": null, "weighting.type": null}, "type": "task"}}, "outputs": [{"data": {"application/vnd.jupyter.widget-view+json": {"model_id": "459927a7aca24d98860e059d5a3ed605", "version_major": 2, "version_minor": 0}, "text/plain": ["GPTaskWidget(lsid='urn:lsid:broad.mit.edu:cancer.software.genepattern.module.analysis:00013')"]}, "metadata": {}, "output_type": "display_data"}], "source": ["knnxvalidation_task = gp.GPTask(genepattern.get_session(0), 'urn:lsid:broad.mit.edu:cancer.software.genepattern.module.analysis:00013')\n", "knnxvalidation_job_spec = knnxvalidation_task.make_job_spec()\n", "knnxvalidation_job_spec.set_parameter(\"data.filename\", \"\")\n", "knnxvalidation_job_spec.set_parameter(\"class.filename\", \"https://datasets.genepattern.org/data/VIB/input.files.list.preprocessed.cls\")\n", "knnxvalidation_job_spec.set_parameter(\"num.features\", \"10\")\n", "knnxvalidation_job_spec.set_parameter(\"feature.selection.statistic\", \"0\")\n", "knnxvalidation_job_spec.set_parameter(\"min.std\", \"\")\n", "knnxvalidation_job_spec.set_parameter(\"num.neighbors\", \"3\")\n", "knnxvalidation_job_spec.set_parameter(\"weighting.type\", \"1\")\n", "knnxvalidation_job_spec.set_parameter(\"distance.measure\", \"1\")\n", "knnxvalidation_job_spec.set_parameter(\"pred.results.file\", \"<data.filename_basename>.pred.odf\")\n", "knnxvalidation_job_spec.set_parameter(\"feature.summary.file\", \"<data.filename_basename>.feat.odf\")\n", "genepattern.display(knnxvalidation_task)"]}, {"cell_type": "markdown", "metadata": {}, "source": ["## 4. View prediction results\n", "\n", "### a. Read the results into a dataframe\n"]}, {"cell_type": "markdown", "metadata": {}, "source": ["<div class=\"alert alert-info\">\n", "<ul>\n", "\t<li>Select the cell containing the job result by clicking anywhere in it.</li>\n", "\t<li>Click on the i icon next to the <code>MergedHTSeqCounts_GSE52778.preprocessed.pred.odf</code> file</li>\n", "\t<li>Select &quot;Send to DataFrame&quot;</li>\n", "\t<li>You will see a new cell that contains 3 lines of code starting with <code>from gp.data import ODF</code></li>\n", "\t<li>Execute this cell</li>\n", "\t<li>You will see the prediction results as a table.</li>\n", "</ul>\n", "</div>\n"]}, {"cell_type": "markdown", "metadata": {}, "source": ["### b. View prediction results"]}, {"cell_type": "markdown", "metadata": {}, "source": ["<div class=\"alert alert-info\">\n", "<h3 style=\"margin-top: 0;\"> Instructions <i class=\"fa fa-info-circle\"></i></h3>\n", "<ol>\n", "    <li>For the <q>prediction results file</q> parameter below, click the down arrow in the file input box.\n", "    </li><li>Right click the <code>MergedHTSeqCounts_GSE52778.preprocessed.pred.odf</code> above and select <q>Copy link address</q>.</li>\n", "    <li>Paste the link into the <q>Prediction Results File</q> parameter.</li>\n", "    <li>Click <strong><em>Run</em></strong>.</li>\n", "    <li>You will see the prediction results in an interactive viewer.</li>\n", "</ol>\n", "</div>\n"]}, {"cell_type": "code", "execution_count": 8, "metadata": {"genepattern": {"description": "Visualizes prediction results in JavaScript", "name": "PredictionResultsViewer", "param_values": {"job.cpuCount": null, "job.memory": null, "job.walltime": null, "prediction.results.file": []}, "type": "task"}}, "outputs": [{"data": {"application/vnd.jupyter.widget-view+json": {"model_id": "239b6a78ea814a9a8e14639b6e641666", "version_major": 2, "version_minor": 0}, "text/plain": ["GPTaskWidget(lsid='urn:lsid:broad.mit.edu:cancer.software.genepattern.module.visualizer:00019')"]}, "metadata": {}, "output_type": "display_data"}], "source": ["predictionresultsviewer_task = gp.GPTask(genepattern.session.get(0), 'urn:lsid:broad.mit.edu:cancer.software.genepattern.module.visualizer:00019')\n", "predictionresultsviewer_job_spec = predictionresultsviewer_task.make_job_spec()\n", "predictionresultsviewer_job_spec.set_parameter(\"prediction.results.file\", \"\")\n", "predictionresultsviewer_job_spec.set_parameter(\"job.memory\", \"2 Gb\")\n", "predictionresultsviewer_job_spec.set_parameter(\"job.walltime\", \"02:00:00\")\n", "predictionresultsviewer_job_spec.set_parameter(\"job.cpuCount\", \"1\")\n", "genepattern.display(predictionresultsviewer_task)"]}, {"cell_type": "markdown", "metadata": {}, "source": ["## References\n", "\n", "Breiman, L., Friedman, J. H., Olshen, R. A., & Stone, C. J. 1984. [Classification and regression trees](https://www.amazon.com/Classification-Regression-Wadsworth-Statistics-Probability/dp/0412048418?ie=UTF8&*Version*=1&*entries*=0). Wadsworth & Brooks/Cole Advanced Books & Software, Monterey, CA.\n", "\n", "Golub, T.R., Slonim, D.K., Tamayo, P., Huard, C., Gaasenbeek, M., Mesirov, J.P., Coller, H., Loh, M., Downing, J.R., Caligiuri, M.A., Bloomfield, C.D., and Lander, E.S. 1999. Molecular Classification of Cancer: Class Discovery and Class Prediction by Gene Expression. [Science 286:531-537](http://science.sciencemag.org/content/286/5439/531.long).\n", "\n", "Lu, J., Getz, G., Miska, E.A., Alvarez-Saavedra, E., Lamb, J., Peck, D., Sweet-Cordero, A., Ebert, B.L., Mak, R.H., Ferrando, A.A, Downing, J.R., Jacks, T., Horvitz, H.R., Golub, T.R. 2005. MicroRNA expression profiles classify human cancers. [Nature 435:834-838](http://www.nature.com/nature/journal/v435/n7043/full/nature03702.html).\n", "\n", "Rifkin, R., Mukherjee, S., Tamayo, P., Ramaswamy, S., Yeang, C-H, Angelo, M., Reich, M., Poggio, T., Lander, E.S., Golub, T.R., Mesirov, J.P. 2003. An Analytical Method for Multiclass Molecular Cancer Classification. [SIAM Review 45(4):706-723](http://epubs.siam.org/doi/abs/10.1137/S0036144502411986).\n", "\n", "Slonim, D.K., Tamayo, P., Mesirov, J.P., Golub, T.R., Lander, E.S. 2000. Class prediction and discovery using gene expression data. In [Proceedings of the Fourth Annual International Conference on Computational Molecular Biology (RECOMB)](http://dl.acm.org/citation.cfm?id=332564). ACM Press, New York. pp. 263-272."]}], "metadata": {"anaconda-cloud": {}, "kernelspec": {"display_name": "Python 3.6", "language": "python", "name": "python3.6"}, "language_info": {"codemirror_mode": {"name": "ipython", "version": 3}, "file_extension": ".py", "mimetype": "text/x-python", "name": "python", "nbconvert_exporter": "python", "pygments_lexer": "ipython3", "version": "3.6.7"}, "toc": {"base_numbering": 1, "nav_menu": {}, "number_sections": false, "sideBar": true, "skip_h1_title": false, "title_cell": "Table of Contents", "title_sidebar": "Contents", "toc_cell": false, "toc_position": {}, "toc_section_display": true, "toc_window_display": false}, "widgets": {"state": {"4f07490870de4bc288ac9aa5a97520c1": {"views": [{"cell_index": 5}]}, "5095e18fe6bc4af99f4a77dec05a3e4d": {"views": [{"cell_index": 3}]}, "f8b01462103f45fea177a69b9e2390f7": {"views": [{"cell_index": 7}]}}, "version": "1.2.0"}, "genepattern": {"repository_url": "http://notebook.genepattern.org/services/sharing/notebooks/359/"}}, "nbformat": 4, "nbformat_minor": 1}